### ISSUE TITLE
3.4-Pytorch-Dataset& Dataloader

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#10]
### Working with data
![image](https://user-images.githubusercontent.com/34474924/195294223-4bb3b9c9-e527-4334-ab98-32ae7e5856f2.png)
要將data先load進去才可以進行training等步驟
`Dataset`: stores the samples and their corresponding labels 對應真實的編號後才好對data進行操作

但以下的解釋比較容易理解
[https://ithelp.ithome.com.tw/articles/10265908](https://ithelp.ithome.com.tw/articles/10265908)
關於機器學習的敘述，我們其實可以發現不管怎樣類型的問題，機器學習的核心都是利用 Data + Answer 來回推 Rules(function) 來進一步做後續的判斷

### 資料集的定義
a collection of data examples and their attributes
一堆收集的例子和他們的屬性/特徵

![image](https://user-images.githubusercontent.com/34474924/211755574-96577564-1cf2-4b41-89c7-0f93f86e811b.png)

`Dataloader`: groups data in batches, enables multiprocessing 提供深度學習的數據型式

PyTorch domain libraries provide a number of pre-loaded datasets (such as FashionMNIST)

Here is an example of how to load the Fashion-MNIST dataset from TorchVision
```python
import torch
from torch.utils.data import Dataset
from torchvision import datasets
from torchvision.transforms import ToTensor
import matplotlib.pyplot as plt


training_data = datasets.FashionMNIST(
    root="data",
    train=True,
    download=True,
    transform=ToTensor()
)

test_data = datasets.FashionMNIST(
    root="data",
    train=False,
    download=True,
    transform=ToTensor()
)
```

`root` is the path where the train/test data is stored,

`train` specifies training or test dataset,

`download=True` downloads the data from the internet if it’s not available at root.

`transform` and `target_transform` specify the feature and label transformations

### Iterating and Visualizing the Datase
We can index Datasets manually like a list: `training_data[index]`. We use `matplotlib` to visualize some samples in our training data.

_____________________________
### Creating a Custom Dataset for your files
A custom Dataset class must implement three functions: `__init__`, `__len__`, and `__getitem__`. 
the FashionMNIST images are stored in a directory img_dir, and their labels are stored separately in a CSV file annotations_file.

In the next sections, we’ll break down what’s happening in each of these functions.

```python
dataset = MyDataset(file)
dataloader = DataLoader(dataset, batch_size, shuffle=True)
//Training: shuffle=True
//Training: shuffle=False
```

```python
class COVID19Dataset(Dataset):
    '''
    x: Features.
    y: Targets, if none, do prediction.
    '''
    def __init__(self, x, y=None):
        if y is None:
            self.y = y
        else:
            self.y = torch.FloatTensor(y)
        self.x = torch.FloatTensor(x)

    def __getitem__(self, idx):
        if self.y is None:
            return self.x[idx]
        else:
            return self.x[idx], self.y[idx]

    def __len__(self):
        return len(self.x)
```
